orttraining/orttraining/test/training_ops/cuda/batch_norm_internal_test.cc

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/framework/tensor.h"
+#include "core/session/inference_session.h"
+#include "test/providers/provider_test_utils.h"
+
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+
+using namespace std;
+
+namespace onnxruntime {
+namespace contrib {
+namespace test {
+
+using namespace onnxruntime::test;
+
+#ifdef USE_CUDA
+static void TestBatchNormInternal(bool test_double = false, bool T_is_half = false,
+                                  bool T1_is_half = false, bool T2_is_half = false,
+                                  const std::vector<int64_t>& input_output_dims = {2, 2, 2, 2}) {
+  OpTester test("BatchNormInternal", 1, kMSDomain);
+  float epsilon = 1e-05f;
+  float momentum = 0.1f;
+  test.AddAttribute("epsilon", epsilon);
+  test.AddAttribute("momentum", momentum);
+
+  std::vector<int64_t> channel_dims{2};
+
+  std::vector<float> X = {-0.2953f, 0.1180f, 1.0973f, -0.1931f, -0.1999f, -0.0237f, 1.5181f, 0.0076f,
+                          -1.0830f, -1.5433f, 0.4327f, -0.9813f, 0.7875f, -0.4080f, -2.3144f, 1.5493f};
+  std::vector<float> scale = {1.0f, 1.0f};
+  std::vector<float> B = {0.0f, 0.0f};
+  std::vector<float> mean = {1.0f, 2.0f};
+  std::vector<float> var = {1.0f, 2.0f};
+
+  std::vector<float> Y = {0.0131f, 0.5210f, 1.7244f, 0.1387f, -0.2708f, -0.1191f, 1.2089f, -0.0922f,
+                          -0.9548f, -1.5203f, 0.9077f, -0.8298f, 0.5796f, -0.4501f, -2.0921f, 1.2358f};
+  std::vector<float> running_mean = {-0.1754f, 0.303106f};
+  std::vector<float> running_var = {0.7812f, 1.5865f};
+  std::vector<float> saved_mean = {-0.306f, 0.114562f};
+  std::vector<float> saved_inv_std = {1.2288f, 0.861317f};
```

Comment:
let's rename this to saved_inv_var to reflect the reality. 
and comment that this test data will only work for CUDA and not CPU.

Reply:
If the cudnn is actually returning saved_inv_std, would this UT also work for CPU impl?

Reply:
It should, but I also infer from the result given in [the calculation](https://github.com/microsoft/onnxruntime/pull/8172#discussion_r674512872) that
- when calculating `saved_inv_std` and `y`, it uses **biased** std/var
- when calculating `running_var`, it uses **unbiased** std/var

As for the [CPU implementation](https://github.com/microsoft/onnxruntime/pull/6946), it always uses the **biased** one for calculation. Not sure why cudnn has such inconsistency itself and which is more reasonable.

And the above difference make the running_var output differ for CPU/CUDA given the same input data.

Reply:
ic... thank you for the detail investigation. 
Could you please also help document this subtle difference in the kernel and UT comment? 

I think the CPU impl is the correct one, as in the ONNX spec, we explicitly mentioned that the variance should be population variance, aka biased variance. 

When the number of sample is large, the difference between biased-var and unbiased-var would be small. Let's note this done and move on.